### DISCUSSION TITLE
Quicksort revisited, Bubble sort fastest of all for small arrays, Quadsort

### DISCUSSION THREAD

**Discussion by dumblob** [MusicTheorist/Rewritten-Grailsort#4] [General]
As promised in https://github.com/daokoder/dao/issues/307#issuecomment-756723644 , I'm copy-pasting the information here.

First of all, thanks for this initiative. It sounds valuable as sorting still seems an unoptimally solved issue in IT. I hope something measurable will come out of it.

2020 seems to have brought some new wind into the topic of sorting. There is e.g. [Quadsort](https://github.com/scandum/quadsort/issues/2 ) - a tight competitor to Grailsort.

But now it comes - a wisely implemented **Bubble sort** for small arrays is the fastest on ILP CPUs (i.e. all CPUs since 2000 including vast majority of embedded ones). When combined with Lomuto and Hoare partitioning schemes for Quick sort, we have a clear speed winner of "all time" - see [Hoare’s Rebuttal and Bubble Sort’s Comeback](https://blog.reverberate.org/2020/05/29/hoares-rebuttal-bubble-sorts-comeback.html ) and the corresponding repo https://github.com/gerben-s/quicksort-blog-post . It can also be made stable with some tweaks (and no additional memory), but there'll be slightly higher number of comparisons.

**Comment by Gaming32** [comment]
Scandum now also has a "Block Merge Sort" implementation based on [Wiki Sort](https://github.com/BonzaiThePenguin/WikiSort). It is called "[Octosort](https://github.com/scandum/octosort)," and its goal is to combine Quadsort with Wiki Sort for optimal performance.

**Comment by Gaming32** [comment]
This has been converted to a discussion for better discussion-y stuff
